kube/src/client/verb.rs

```diff
@@ -0,0 +1,346 @@
+}
+type Result<T, E = Error> = std::result::Result<T, E>;
+
+/// An action that Kube can take
+pub trait Verb {
+    /// Decodes the response given from the server
+    /// Will typically be [`DecodeSingle`]
+    type ResponseDecoder: TryFuture + From<Response<Body>>;
+
+    /// Prepare a HTTP request that takes the action
+    ///
+    /// Should include request-specific options, but not global options (such as the base URI or authentication tokens)
+    fn to_http_request(&self) -> Result<Request<Body>>;
+}
+
+/// Get a single object
+pub struct Get<'a, Kind: Resource, Scope> {
+    /// The name of the object
+    pub name: &'a str,
+    /// The scope that the object will be queried from
+    pub scope: &'a Scope,
+    /// The type of the object
+    pub dyn_type: &'a Kind::DynamicType,
+}
+impl<'a, Kind: Resource + DeserializeOwned, Scope: NativeScope<Kind>> Verb for Get<'a, Kind, Scope> {
+    type ResponseDecoder = DecodeSingle<Kind>;
+
+    fn to_http_request(&self) -> Result<Request<Body>> {
+        Request::get(format!(
+            "{}/{}",
+            Kind::url_path(&self.dyn_type, self.scope.namespace()),
+            self.name
+        ))
+        .body(Body::empty())
+        .context(BuildRequestFailed)
+    }
+}
+
+/// List all objects of a resource type
+pub struct List<'a, Kind: Resource, Scope> {
+    /// The scope that the objects will be queried from
+    pub scope: &'a Scope,
+    /// The type of the objects
+    pub dyn_type: &'a Kind::DynamicType,
+    /// The query to filter the objects by
+    pub query: &'a Query<'a>,
+
+    /// Limit the number of results.
+    ///
+    /// If there are more results, the server will respond with a continue token which can be used to fetch another page
+    /// of results. See the [Kubernetes API docs](https://kubernetes.io/docs/reference/using-api/api-concepts/#retrieving-large-results-sets-in-chunks)
+    /// for pagination details.
+    pub limit: Option<u32>,
+
+    /// Fetch a second page of results.
+    ///
+    /// After listing results with a `limit`, a continue token can be used to fetch another page of results.
+    pub continue_token: Option<&'a str>,
+}
+impl<'a, Kind: Resource + DeserializeOwned, Scope: scope::Scope> Verb for List<'a, Kind, Scope> {
+    type ResponseDecoder = DecodeSingle<ObjectList<Kind>>;
+
+    fn to_http_request(&self) -> Result<Request<Body>> {
+        let mut url = format!("{}?", Kind::url_path(&self.dyn_type, self.scope.namespace()));
+        let mut qp = form_urlencoded::Serializer::new(&mut url);
+        self.query.populate_qp(&mut qp);
+        if let Some(limit) = self.limit {
+            qp.append_pair("limit", &limit.to_string());
+        }
+        if let Some(cont) = self.continue_token {
+            qp.append_pair("continue", cont);
+        }
+        Request::get(url).body(Body::empty()).context(BuildRequestFailed)
+    }
+}
+/// Common query parameters used to select multiple objects
+#[derive(Default)]
+pub struct Query<'a> {
```

Comment:
This does mean we take a sharp break from apimachinery where we literally transcribe the `ListOptional`, and `DeleteOptional` into `ListParams` and `DeleteParams`. Now we are factoring out bits of that type, when even [apimachinery doesn't do that](https://github.com/kubernetes/apimachinery/blob/57f2a0733447cfd41294477d833cce6580faaca3/pkg/apis/meta/v1/types.go#L335).

Reply:
True, but the current approach doesn't really translate well anyway (such as the way that we ignore just ignore some `ListParams` fields on watch, and so on).